plugins/modules/aci_esg_ip_subnet_selector.py

```diff
@@ -0,0 +1,303 @@
+                "name": "production"
+            }
+        }
+    }
+filter_string:
+  description: The filter string used for the request
+  returned: failure or debug
+  type: str
+  sample: ?rsp-prop-include=config-only
+method:
+  description: The HTTP method used for the request to the APIC
+  returned: failure or debug
+  type: str
+  sample: POST
+response:
+  description: The HTTP response from the APIC
+  returned: failure or debug
+  type: str
+  sample: OK (30 bytes)
+status:
+  description: The HTTP status from the APIC
+  returned: failure or debug
+  type: int
+  sample: 200
+url:
+  description: The HTTP url used for the request to the APIC
+  returned: failure or debug
+  type: str
+  sample: https://10.11.12.13/api/mo/uni/tn-production.json
+"""
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.cisco.aci.plugins.module_utils.aci import (
+    ACIModule,
+    aci_argument_spec,
+)
+
+
+def main():
+    argument_spec = aci_argument_spec()
+    argument_spec.update(
+        tenant=dict(type="str", aliases=["tenant_name"]),
+        ap=dict(type="str", aliases=["app_profile", "app_profile_name"]),
+        esg=dict(type="str", aliases=["esg_name"]),
+        ip=dict(type="str", aliases=["subnet_ip"]),
+        description=dict(type="str", aliases=["ip_subnet_selector_description"]),
+        state=dict(
+            type="str",
+            default="present",
+            choices=["absent", "present", "query"],
+        ),
+        name_alias=dict(type="str"),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ["state", "absent", ["tenant", "ap", "esg", "ip"]],
+            ["state", "present", ["tenant", "ap", "esg", "ip"]],
+        ],
+    )
+
+    aci = ACIModule(module)
+    tenant = module.params.get("tenant")
+    ap = module.params.get("ap")
+    esg = module.params.get("esg")
+    ip = module.params.get("ip")
+    description = module.params.get("description")
+    state = module.params.get("state")
+
+    match_expression = "ip=='{0}'".format(ip)
+    aci_rn = "epselector-[{0}]".format(match_expression)
+    dn = "uni/tn-{0}/ap-{1}/esg-{2}/{3}".format(tenant, ap, esg, aci_rn)
+    aci.construct_url(
+        root_class=dict(
+            aci_class="fvTenant",
+            aci_rn="tn-{0}".format(tenant),
+            module_object=tenant,
+            target_filter={"name": tenant},
+        ),
+        subclass_1=dict(
+            aci_class="fvAp",
+            aci_rn="ap-{0}".format(ap),
+            module_object=ap,
+            target_filter={"name": ap},
+        ),
+        subclass_2=dict(
+            aci_class="fvESg",
+            aci_rn="esg-{0}".format(esg),
+            module_object=esg,
+            target_filter={"name": esg},
+        ),
+        subclass_3=dict(
+            aci_class="fvEPSelector",
+            aci_rn=aci_rn,
+            module_object=dn,
```

Comment:
I would put ip here and not dn. This need to be null when ip is not passed in order to make a class query.
I don't think you need to build the dn.